### PR TITLE
Input Plugin for Year

### PR DESCRIPTION
**Pull request by rachelnicole**
added linting and proper generator input plugin structure. still need tests.

### CODE REVIEW

**Comment by rachelnicole** [comment]
Realizing that the dom js has ES2015 stuff in it. I'm assuming I should put some linting flags and change it back so we don't have to shim?

**Comment by rachelnicole** [comment]
Tests and Linting are now complete.

File: .eslintrc.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+extends:
+  - punchcard
+
+env:
```
Comment (Snugug) [inline .eslintrc.yml]: The linting doesn't follow the new linting pattern. Please see https://github.com/punchcard-cms/input-plugin-time for the new linting pattern

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+language: node_js
+node_js:
+  - '5.7'
+
+# Setup
+cache:
+  directories:
+    - node_modules
```
Comment (Snugug) [inline .travis.yml]: Missing Code Coverage `after_success`

File: index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        minYear: '1900',
+        maxYear: '2016',
+      },
+      type: 'select',
+      settings: {
+        empty: true,
+      },
+    },
+  },
+  html: '<label for="{{year.id}}">{{year.label}}</label><select id="{{year.id}}" name="{{year.name}}"><option value="{{year.value}}"  class="year-range"></option></select></label>',
```
Comment (Snugug) [inline index.js]: What is the goal of this input plugin? To generate all years between the `min` and `max` year? This HTML won't generate that. This also won't generate a `selected` option, which we need.

@scottnath generated content in input plugins isn't something we've talked about. Thoughts on this for a select?

File: lib/yearRange.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Function to generate a span of years based on minimum and maximum parameters
+ *
+ * @param {Object} input - inputs in this plugin's instance and their values as {String}
+ * @param {Object} minYear - inputs minimum year value {String}
+ * @param {Object} maxYear - inputs maxmium year value {String}
+ * *
+ * @module yearRange
+ */
+
+module.exports = function yearRange(input, minYear, maxYear) {
```
Comment (Snugug) [inline lib/yearRange.js]: As we talked about, UI scripts will have `inputs` and `settings` passed to it, and only those two things. `minYear` and `maxYear` are not things that will be passed to this function

File: lib/yearRange.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  var select = input.querySelector('.year-range');
+
+  for (var i = minYear; i <= maxYear; i++) {
```
Comment (Snugug) [inline lib/yearRange.js]: It looks like options are being generated entirely client side. We need these to be generated in HTML, not through JS

File: package.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  "version": "0.0.1",
+  "description": "A simple select input for years based on a range",
+  "main": "index.js",
+  "scripts": {
+    "test": "npm run nyc",
+    "pretest": "npm run lint",
+    "ava": "ava | tap-diff",
+    "ava:watch": "ava --watch | tap-diff",
+    "nyc": "nyc --all npm run ava",
+    "lint": "eslint index.js lib && eslint -c ./.eslintrc-ava.yml tests",
```
Comment (Snugug) [inline package.json:12]: Don't have an `.eslintrc-ava.yml` file

**Comment by Snugug** [comment]
@rachelnicole this needs some work. Please take a look and update

File: index.js (hunk 3 above)
Comment (scottnath) [inline index.js]: @Snugug I think generated content in input plugins is def something we need, and this scenario is probably the easiest and most direct way to figure out how to do that.

`selected` can be generated, but the code needs to be added. See: https://github.com/punchcard-cms/input-plugin-select/blob/master/index.js#L48 for how to do `selected`.

Another option might be to _not_ use a separate javascript, but instead put the js into the `html` attribute? Or **better yet**, maybe do it with nunjucks??

File: lib/yearRange.js (hunk 5 above)
Comment (scottnath) [inline lib/yearRange.js]: yup...I think this brings us to instead using a nunjucks solution. You could move minYear/maxYear into settings - iterate over them to create `<options>`

File: package.json (hunk 6 above)
Comment (rachelnicole) [inline package.json:12]: It doesn't? I see it here: https://github.com/rachelnicole/input-plugin-year/blob/e6131becb8de621238e51e1ccb8d4dd4fef67270/.eslintrc-ava.yml

**Comment by scottnath** [comment]
@rachelnicole were you able to attempt doing this in the html with nunjucks?

**Comment by rachelnicole** [comment]
@scottnath i haven't even tried yet, i was fixing all the other PR's and been in meetings.

**Comment by Snugug** [comment]
@scottnath @rachelnicole question: why are we creating a year plugin?

There is no "year" input type like there is for month or date or datetime. If I wanted a "year" input for my site, why wouldn't I just configure a `select` with the options I'd like, or a `number` input with the appropriate min/max? Why do I need a pre-configured select or number input to do year for me?

Especially as I would still need to configure this

**Comment by scottnath** [comment]
If you did it with `number`, then that would be an input.
If you did it with `select`, you'd have to add each year, or create a js to loop through min/max to create each `option`. The current `select` iterates over existing data.

Year in a `select` is a very common input type. It was also supposed to help with testing the use of a ux javascript.
however, on the ux-js point, that is not the case anymore as the iteration over min/max can and should be done inside the html with nunjucks

File: lib/yearRange.js (hunk 5 above)
Comment (rachelnicole) [inline lib/yearRange.js]: working on this now ✨

Comment (rachelnicole) [inline lib/yearRange.js]: Just found out that nunjucks has a range setting, which I was unaware of. This seems perfect for our use case. I'm wondering if we add the range min and max to settings, if we can put nested values inside the nunjucks range function like 

```
{% for i in range({{settings.min}}, {{settings.max}}) -%}
  <option value="{{ i }}">{{ i }}</option>,
{%- endfor %}
```

 You can see it here:

https://mozilla.github.io/nunjucks/templating.html#range-start-stop-step

File: index.js [hunk 7]
```diff
@@ -31,15 +27,12 @@ module.exports = {
...
       label: 'Please choose a year',
       placeholder: 'year',
-      options: {
-        minYear: '1900',
-        maxYear: '2016',
-      },
       type: 'select',
       settings: {
-        empty: true,
+        minYear: 2000,
```
Comment (Snugug) [inline index.js:32]: Change to `min`

File: index.js [hunk 8]
```diff
@@ -31,15 +27,12 @@ module.exports = {
...
+        maxYear: 2016,
```
Comment (Snugug) [inline index.js:33]: Change to `max`

File: index.js [hunk 9]
```diff
@@ -31,15 +27,12 @@ module.exports = {
...
       },
     },
   },
-  html: '<label for="{{year.id}}">{{year.label}}</label><select id="{{year.id}}" name="{{year.name}}"><option value="{{year.value}}"  class="year-range"></option></select></label>',
+  html: '<label for="{{year.id}}">{{year.label}}</label><select id="{{year.id}}" name="{{year.name}}">{% for i in range(year.settings.minYear, year.settings.maxYear) -%}<option value="{{ i }}">{{ i }}</option>{%- endfor %}</select></label>',
```
Comment (Snugug) [inline index.js]: Need check to see if `year.value` is equal to `i` to set the `selected` attribute on the `option` tag

File: tests/plugin.js [hunk 10]
```diff
@@ -1,6 +1,34 @@
...
+    label: 'Please choose a year',
+    placeholder: 'year',
+    type: 'select',
+    settings: {
+      minYear: 2000,
+      maxYear: 2016,
+    },
+  },
+};
+
```
Comment (Snugug) [inline tests/plugin.js:22]: Add test for `selected`

File: lib/validation.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {Object} [settings.target] - the triggering input's settings as an {Object}
+ * @param {Object} [settings.all] - all settings in this plugin instance as an {Object}
+ * @param {Object} [settings.all.year] - settings of input year {Object}
+ *
+ * @returns {Bool|String} true or a string with the text describing the error
+ *
+ * @module yearValidation
+ */
+
+module.exports = function yearValidation(input, settings) {
```
Comment (Snugug) [inline lib/validation.js:22]: Make sure you update `validation` to make sure that the selected year falls within the year range defined in `settings`

File: package.json (hunk 6 above)
Comment (Snugug) [inline package.json:12]: I see it now

File: tests/validation.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  target: {
+    empty: false,
+  },
+  all: {
+    select: {
+      empty: false,
+    },
+  },
+};
+
```
Comment (Snugug) [inline tests/validation.js:42]: Add validation test to make sure it's within range

**Comment by rachelnicole** [comment]
@Snugug added validation test!

File: lib/validation.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  console.log(input);
+  console.log('-----');
+  console.log(settings);
```
Comment (scottnath) [inline lib/validation.js]: remove `console.log`s please.

File: tests/plugin.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      minYear: 2000,
+      maxYear: 2016,
+    },
+    value: 2006,
+  },
+};
+
+test('select fills options', t => {
+  const rendered = nunjucks.renderString(plugin.html, data);
+  t.true(contains(rendered, '<option value=\"2005\" >2005</option>'), 'Select must contain 2005');
```
Comment (scottnath) [inline tests/plugin.js:25]: can you please add a couple `t.false` tests for dates outside min/max range?

Comment (rachelnicole) [inline tests/plugin.js:25]: In addition to the check for dates outside the range that are inside validation?

**Comment by rachelnicole** [comment]
@scottnath your wish for more range tests has been granted 🔮✨

**Pull request merged**